### ISSUE TITLE
is seconds_frozen always taken into account?

### ISSUE THREAD

**Issue by wsw70** [niccokunzmann/hanging_threads#25]
I have a program with multiple threads, each of them being a `while True: (...) time.sleep(x)` loop. I have set `seconds_frozen` to a value higher than the longest `sleep` but still get dumps from `hanging_threads`,  for example:

```
--------------------Thread 139845980632832--------------------
  File "/usr/lib/python3.5/threading.py", line 882, in _bootstrap
        self._bootstrap_inner()
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
        self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
        self._target(*self._args, **self._kwargs)
  File "webserver.py", line 139, in check_source_changes
        suspend_for("10s")
  File "webserver.py", line 151, in suspend_for
        time.sleep(sleeptime)
```

In the case above, `sleeptime` is `10` and `hanging_threads` was initialized with

```
monitoring_thread = start_monitoring(seconds_frozen=700, tests_per_second=1)
```

Is there something else I should do?

**Comment by niccokunzmann** [comment]
Hi, can you show the code?

If there is something like

    while 1:
        time.sleep(...)

I can understand that behavior as it looks  every second (in your case) if the thread has the same stacktrace.

Also, in #15 @coffenbacher proposed that nobody likes to see the `time.sleep` as it is not really a deadlock. We want to make it possible to remove the `time.sleep` and other calls. If you like, you can propose how you like to modify the module or help us implementing it.

How would you like to go on? What is the behavior you expect?

**Comment by wsw70** [comment]
@niccokunzmann : I am not sure if I got the concept of the module right, but I expected that the code below:

```
import threading
import hanging_threads
import time

hanging_threads.start_monitoring(seconds_frozen=10, tests_per_second=1)

def worker():
    while True:
        print('hello')
        time.sleep(5)

threading.Thread(target=worker).start()
```

would monitor if the thread has not been inactive for 10 seconds (`seconds_frozen=10`), checking once every second (`tests_per_second=1`, which seems to be the minimum possible rate (?)).

What I get when running it is

```
hello
hello
hello
--------------------    Thread 15088    --------------------
  File "C:\Python35\lib\threading.py", line 882, in _bootstrap
	self._bootstrap_inner()
  File "C:\Python35\lib\threading.py", line 914, in _bootstrap_inner
	self.run()
  File "C:\Python35\lib\threading.py", line 862, in run
	self._target(*self._args, **self._kwargs)
  File "C:/Users/yop/.PyCharmCE2016.3/config/scratches/scratch_1.py", line 10, in worker
	time.sleep(5)
--------------------    Thread 22540    --------------------
  File "C:\Python35\lib\threading.py", line 1288, in _shutdown
	t.join()
  File "C:\Python35\lib\threading.py", line 1054, in join
	self._wait_for_tstate_lock()
  File "C:\Python35\lib\threading.py", line 1070, in _wait_for_tstate_lock
	elif lock.acquire(block, timeout):
hello
```

--> which is a behaviour I do not understand: the thread was active every 5 seconds, but was still reported as inactive.

**Comment by niccokunzmann** [comment]
Thanks for the code. So: the reason is: it checks every second if the thread is at the same location and it is:

    time.sleep(5)

If you do

    time.sleep(2)
    time.sleep(3)

The message should go away.

To implementation: If we check if the loop continues/the thread stack changes, we would need to get notified by Python if interpretation goes on. This may require (I believe) executing Python code on every step which slows down the program, changes the execution of the existing threads and changes their behavior, eventually removing race conditions and deadlocks. So, the test-interval-aproach is less changing.

You can also increase the checking and hope that once in a while the check notices that the print is executed. `tests_per_second=1000` Maybe that changes something.

You showed a false-positive (wrongly detected hanging thread). We can document this as a problem.

What are your thoughts about this?

**Comment by artemrizhov** [comment]
@niccokunzmann Is there any information in the stack trace that can be used to avoid such problems? Maybe it provides some time stamp or unique id of the frame?

**Comment by artemrizhov** [comment]
@wsw70 In any case the problem should be reduced when #21 and #31 are implemented.

**Comment by niccokunzmann** [comment]
Maybe there is: f.f_trace - tracing function for this frame, or None [source](https://docs.python.org/2/library/inspect.html) 

So: We can create a kind of monitoring where we check if an instruction is executed.

If I am concerned about performance: It can be switched on when we assume a thread hangs and then we see if it is such a loop.
What is to be discussed: 

- A while loop can make a GUI unresponsive. This would also be considered a hanging thread.
- By default, I would rather print false threads and too many than leave out possibly correct ones: So I would vote for adding options like `check_frame_execution_after_seconds` to switch on such behavior.

@artemrizhov @wsw70 What are your thoughts?

**Comment by artemrizhov** [comment]
> A while loop can make a GUI unresponsive. This would also be considered a hanging thread.

@niccokunzmann Good point. If so, I propose to leave this as is. Once possible hanging thread is detected, developer should be smart enough to understand if he is interesting in this certain case or not. False reports should not be a big problem once #21 and #31 are implemented. If this will be a problem afterwards then I'd like to get some explanation why this is a problem to help us find the best solution. For now it seems the main problem is the repeating messages.

Btw, I'm still thinking on #31. Probably some examples of real problems are needed to find best solution. Hopefully closing #21 will help us understand where we go next.

**Comment by artemrizhov** [comment]
@niccokunzmann I think the issue label should be changed from "bug" to "question", because this is not a bug but a feature. :)

@wsw70 New version v2.0.0 contains feature #21. Could you please try it and tell us how it works for you?

**Comment by remdragon** [comment]
There's a somewhat simple but dirty way to fix this problem.

All you have to do is monkey-patch time.sleep() to flag that thread as unhangable and when the original time.sleep returns, you reset the thread's status, something like this:

```
sleeping_threads = {}
time_sleep = time.sleep
def start_monitoring ( ... ):
    def new_sleep ( seconds ):
        global sleeping_threads
        id = threading.current_thread().ident
        sleeping_threads[id] = True # prevents this thread id from being flagged as hung
        try:
            time_sleep ( seconds )
        finally:
            del sleeping_threads[id]
    time.sleep = new_sleep
    ...
```
